be/src/exec/spill/spiller.cpp

```diff
@@ -0,0 +1,230 @@
+    DCHECK_GT(serialize_sz, 4);
+
+    auto* buff = reinterpret_cast<uint8_t*>(context.io_buffer.data());
+    UNALIGNED_STORE64(buff, serialize_sz);
+    buff += sizeof(serialize_sz);
+
+    for (const auto& column : chunk->columns()) {
+        buff = serde::ColumnArraySerde::serialize(*column, buff);
+    }
+
+    RETURN_IF_ERROR(writable->append(context.io_buffer));
+    return Status::OK();
+}
+
+StatusOr<ChunkUniquePtr> ColumnSpillFormater::restore_from_fmt(SpillFormatContext& context,
+                                                               std::unique_ptr<RawInputStreamWrapper>& readable) const {
+    size_t serialize_sz;
+    RETURN_IF_ERROR(readable->read_fully(&serialize_sz, sizeof(serialize_sz)));
+    DCHECK_GT(serialize_sz, sizeof(serialize_sz));
+    context.io_buffer.resize(serialize_sz);
+    auto buff = reinterpret_cast<uint8_t*>(context.io_buffer.data());
+    RETURN_IF_ERROR(readable->read_fully(buff, serialize_sz - sizeof(serialize_sz)));
+
+    auto chunk = _chunk_builder();
+    const uint8_t* read_cursor = buff;
+    for (const auto& column : chunk->columns()) {
+        read_cursor = serde::ColumnArraySerde::deserialize(read_cursor, column.get());
+    }
+    return chunk;
+}
+
+Status ColumnSpillFormater::flush(std::unique_ptr<WritableFile>& writable) const {
+    // TODO: test flush async and sync
+    RETURN_IF_ERROR(writable->flush(WritableFile::FLUSH_ASYNC));
+    return Status::OK();
+}
+
+StatusOr<std::unique_ptr<SpillFormater>> SpillFormater::create(SpillFormaterType type, ChunkBuilder chunk_builder) {
+    if (type == SpillFormaterType::SPILL_BY_COLUMN) {
+        return std::make_unique<ColumnSpillFormater>(std::move(chunk_builder));
+    } else {
+        return Status::InternalError(fmt::format("unsupported spill type:{}", type));
+    }
+}
+
+Status Spiller::prepare(RuntimeState* state) {
+    // prepare
+    ASSIGN_OR_RETURN(_spill_fmt, SpillFormater::create(_opts.spill_type, _opts.chunk_builder));
+
+    for (size_t i = 0; i < _opts.mem_table_pool_size; ++i) {
+        // TODO: init mem table
+        return Status::NotSupported("TODO: implements MemTable");
+    }
+
+    _file_group = std::make_shared<SpilledFileGroup>(*_spill_fmt);
+
+    return Status::OK();
+}
+
+Status Spiller::_open(RuntimeState* state) {
+    std::lock_guard guard(_mutex);
+    if (_has_opened) {
+        return Status::OK();
+    }
+
+    // init path provider
+    ASSIGN_OR_RETURN(_path_provider, _opts.path_provider_factory());
+    RETURN_IF_ERROR(_path_provider->open(state));
+    _has_opened = true;
+
+    return Status::OK();
+}
+
+Status Spiller::_flush_and_closed(std::unique_ptr<WritableFile>& writable) {
+    // flush
+    RETURN_IF_ERROR(_spill_fmt->flush(writable));
+    // be careful close method return a status
+    RETURN_IF_ERROR(writable->close());
+    writable.reset();
+    return Status::OK();
+}
+
+Status Spiller::_run_flush_task(RuntimeState* state, const MemTablePtr& mem_table) {
```

Comment:
if query is cancelled, should skip

Reply:
it should be a enhancement. I will do it later